### DISCUSSION TITLE
HasManyThrough relation media doesn't remove media.

### DISCUSSION THREAD

**Discussion by shankhadevpadam** [spatie/laravel-medialibrary#2659] [Q&A]
I have 3 tables

```
users
     - id
     - name
 ```
```
user_packages
    - id
    - user_id
    - package_id
```
```
user_package_travellers
    - id
    - user_package_id
 ```
And I stored media for the user and user_package_traveller model. But the problem is when deleting the user it doesn't delete the media associated with package travellers but delete media associated with the user.

```php
// User.php
public function userPackageTraveller()
{
    return $this->hasManyThrough(
        UserPackageTraveller::class,
        UserPackage::class,
        'user_id',
        'user_package_id'
    );
}
```

**Comment by freekmurze** [comment]
Could you create a failing test for this for me to fix?

**Reply by shankhadevpadam** [reply]
Not only hasManyThrough relation it also not deleting the hasOne or hasMany relation.

for eg.

```
users
  - id
  - name
  - email
```
```
reviews
  - id
  - user_id
  - title
  - publish_date
```

```php
// User.php

public function reviews()
{
	return $this->hasMany(Review::class);
}

```
and I have media associated with the review but when I delete the user it deletes the user and associated review but doesn't delete media associated with the review.

**Reply by erikn69** [reply]
are you using cascade delete on db? or deleting method on boot?
https://github.com/spatie/laravel-medialibrary/blob/a7e961d75ffc0547870c23c2e9a8cd334f2a20a4/src/InteractsWithMedia.php#L42-L57

**Reply by shankhadevpadam** [reply]
I used cascade delete on db.

**Reply by erikn69** [reply]
That's is why 😄
The package can´t know when there is a `deleting` from out of laravel, you have to make a delete from boot method and everything works

```php
 public static function boot() 
 { 
     parent::boot();
     static::deleting(function ($model) { 
          // delete relations
     });
 }
```
Read laravel's documentation

**Reply by shankhadevpadam** [reply]
Thanks for answering.

**Reply by shankhadevpadam** [reply]
Deleting relation also does not work only worked when the associated model is deleted.

```php
// This work fine
foreach($user->reviews as $review) {
    $review->delete();
}
```
```php
// This doesn't work
$user->reviews()->delete();
```

**Reply by erikn69** [reply]
Looks like you don't know how Laravel works,
Yes, bulk delete does not trigger any events

**Reply by shankhadevpadam** [reply]
I managed myself.

```php
public static function boot()
{
    parent::boot();

    self::deleting(function($user) {
        $user->reviews()->each(function($review) {
            $review->delete();
        });
    });
}
```
